posts/2021/08/github-actions-benchmarks.md

```diff
@@ -0,0 +1,369 @@
+
+# Can we make it run faster without losing accuracy?
+
+We just found out that this approach is sensitive enough _but_ it takes two hours to run. Is it possible
+to reduce that time without sacrificing sensitivity? We need to remember that `asv` is running several
+passes and repeats to reduce the measurement error, but maybe some of those default counter-measures
+are not needed. Namely:
+
+- The benchmark runs with `--interleave-processes`, but it can be disabled with `--no-interleave-processes`.
+The help message for this flag says:
+
+    > Interleave benchmarks with multiple processes across commits. This can avoid measurement biases
+    > from commit ordering, can take longer.
+
+    How much longer? Does it help keep error under control? We should measure that.
+
+- By default, all tests are run several times, with different schedules. There are two
+[benchmark attributes](https://asv.readthedocs.io/en/stable/benchmarks.html#timing-benchmarks)
+that govern these settings: `processes` and `repeat`. `processes` defaults to `2`, which means that
+the full suite will be run twice per commit. If we only do one pass (`processes=1`), we will reduce
+the running time in half, but will we lose too much accuracy?
+
+To answer both questions, we added more entries to the data collection workflow shown above by
+[parameterizing the `asv` command-line options](https://github.com/jaimergp/scikit-image/blob/e561996/.github/workflows/benchmarks-cron.yml#L14-L23).
+
+These are the results!
+
+## No process interleaving
+
+Disabling process interleaving should be faster and maybe the accuracy loss is not that bad. But...
+how bad? Here are the results:
+
+![Reliability of benchmarks in GitHub Actions, no interleaving. This 2D plots shows a 16-day timeseries in the X axis.
+  Each data point in the X axis corresponds to a cloud of 75 measurements (one per benchmark test).
+  The y-axis spread of each cloud corresponds to the performance ratio. Ideal measurements would have
+  a performance ratio of 1.0, since both runs returned the exact same performance. In practice this
+  does not happen. With no interleaving, the vertical spread is more evident, with several clouds
+  spreading beyond the desired interval. One particular outlier happened on the first Saturday,
+  with half the cloud below 0.75.](/images/2021/08/github-actions-benchmark-no-interleaving.png)
+
+* Average time taken: 1h39min
+* Minimum and maximum ratios observed: 0.43, 1.5
+* Mean and standard deviation of ratios: 0.99, 0.07
+* Proportion of false positives: 6/66 = 9.99%
+
+## Single-pass with `processes=1`
+
+In this configuration, we expect a drastic 50% running time reduction, since we will only do one pass
+per commit, instead of two. However, the accuracy loss might be too dramatic... Let's see!
+
+![Reliability of benchmarks in GitHub Actions, single pass. This 2D plots shows a 16-day timeseries in the X axis.
```

Comment:
For the same reason as mentioned in the third alt text, I'd recommend moving the main takeaway earlier in the alt text like 

```
Reliability of benchmarks in GitHub Actions, single pass. With a single process, 
the y-spread is significantly wider. Some data points can
be observed even beyond a ratio of 2.5 (it should be 1.0!), and the overall visual spread is
larger; i.e. the clouds are taller than in the ideal case of having a small, contained cloud at 1.0.
This 2D plots shows a 16-day timeseries in the X axis.
Each data point in the X axis corresponds to a cloud of 75 measurements (one per benchmark test).
The y-axis spread of each cloud corresponds to the performance ratio. Ideal measurements would have
a performance ratio of 1.0, since both runs returned the exact same performance. In practice this
does not happen. 


```